Address safer CPP failures in WebFoundTextRangeController.cpp

pxlcoder opened the pull request:
#### ae1342640d59b646fdf527543ff2ee038b95f68c
<pre>
Address safer CPP failures in WebFoundTextRangeController.cpp
<a href="https://bugs.webkit.org/show_bug.cgi?id=290030">https://bugs.webkit.org/show_bug.cgi?id=290030</a>
<a href="https://rdar.apple.com/147385786">rdar://147385786</a>

Reviewed by Rupin Mittal.

* Source/WebCore/page/PageOverlay.cpp:
(WebCore::PageOverlay::layer const):
(WebCore::PageOverlay::protectedLayer const):
(WebCore::PageOverlay::layer): Deleted.
* Source/WebCore/page/PageOverlay.h:

Drive-by: Make getters const.

* Source/WebCore/page/PageOverlayController.cpp:
(WebCore::PageOverlayController::layerForOverlay const):
* Source/WebCore/page/PageOverlayController.h:
* Source/WebKit/SaferCPPExpectations/UncheckedCallArgsCheckerExpectations:
* Source/WebKit/SaferCPPExpectations/UncountedCallArgsCheckerExpectations:
* Source/WebKit/WebProcess/WebPage/WebFoundTextRangeController.cpp:
(WebKit::WebFoundTextRangeController::findTextRangesForStringMatches):
(WebKit::WebFoundTextRangeController::replaceFoundTextRangeWithString):
(WebKit::WebFoundTextRangeController::decorateTextRangeWithStyle):
(WebKit::WebFoundTextRangeController::clearAllDecoratedFoundText):
(WebKit::WebFoundTextRangeController::didBeginTextSearchOperation):
(WebKit::WebFoundTextRangeController::addLayerForFindOverlay):
(WebKit::WebFoundTextRangeController::removeLayerForFindOverlay):
(WebKit::WebFoundTextRangeController::createTextIndicatorForRange):
(WebKit::WebFoundTextRangeController::flashTextIndicatorAndUpdateSelectionWithRange):
(WebKit::WebFoundTextRangeController::flashTextIndicatorAndUpdateSelectionWithPDFRange):
(WebKit::WebFoundTextRangeController::rectsForTextMatchesInRect):
(WebKit::WebFoundTextRangeController::frameForFoundTextRange const):

Canonical link: <a href="https://commits.webkit.org/292430@main">https://commits.webkit.org/292430@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/69265b2773f9ad9e1f8d0adb4a9ea6484e9f711c

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/95920 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/15534 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/5479 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/100979 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/46429 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/97965 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/15829 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/23966 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/73137 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/30357 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/98923 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/11850 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/86643 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/53461 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/11584 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/4392 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/45764 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/81752 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/4494 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/103008 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/22987 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/16757 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/82192 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/23238 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/82667 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/81554 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/26127 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/3577 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/16324 "Built successfully") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/15452 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/22950 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/28105 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/22609 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/26089 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/24350 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

RupinMittal commented on Source/WebKit/WebProcess/WebPage/WebFoundTextRangeController.cpp:
```diff
@@ -95,7 +95,7 @@ void WebFoundTextRangeController::findTextRangesForStringMatches(const String& s
 
 #if ENABLE(PDF_PLUGIN)
-    for (RefPtr frame = &m_webPage->corePage()->mainFrame(); frame; frame = frame->tree().traverseNext()) {
+    for (RefPtr frame = &m_webPage->protectedCorePage()->mainFrame(); frame; frame = frame->tree().traverseNext()) {
```
`mainFrame()` is a trivial function so I believe we don't need to use `protectedCorePage()` here. Just `corePage()` as-is should be ok.

RupinMittal commented on Source/WebKit/WebProcess/WebPage/WebFoundTextRangeController.cpp:
```diff
@@ -225,44 +225,46 @@ void WebFoundTextRangeController::clearAllDecoratedFoundText()
-        m_findPageOverlay = WebCore::PageOverlay::create(*this, WebCore::PageOverlay::OverlayType::Document);
-        m_webPage->corePage()->pageOverlayController().installPageOverlay(*m_findPageOverlay, WebCore::PageOverlay::FadeMode::Fade);
+        RefPtr findPageOverlay = WebCore::PageOverlay::create(*this, WebCore::PageOverlay::OverlayType::Document);
+        m_webPage->protectedCorePage()->pageOverlayController().installPageOverlay(*m_findPageOverlay, WebCore::PageOverlay::FadeMode::Fade);
```
Given what the old code was doing, it seems like we want to use `findPageOverlay` here instead of `m_findPageOverlay`:

`m_webPage->protectedCorePage()->pageOverlayController().installPageOverlay(*findPageOverlay, WebCore::PageOverlay::FadeMode::Fade);`

RupinMittal replied:
Actually even better might be to use a protected function so we don't need to the `WTFMove` and` RefPtr { }` below:

`else {`
`       m_findPageOverlay = WebCore::PageOverlay::create(*this, WebCore::PageOverlay::OverlayType::Document);`
`       m_webPage->corePage()->pageOverlayController().installPageOverlay(*protectedFindPageOverlay(), WebCore::PageOverlay::FadeMode::Fade);`
`   }`

`   protectedFindPageOverlay()->setNeedsDisplay();`

pxlcoder replied:
Added the protected function, but still using `WTFMove` to avoid an unnecessary ref.

RupinMittal commented on Source/WebKit/WebProcess/WebPage/WebFoundTextRangeController.cpp:
```diff
@@ -225,44 +225,46 @@ void WebFoundTextRangeController::clearAllDecoratedFoundText()
+        RefPtr findPageOverlay = WebCore::PageOverlay::create(*this, WebCore::PageOverlay::OverlayType::Document, WebCore::PageOverlay::AlwaysTileOverlayLayer::Yes);
+        m_webPage->protectedCorePage()->pageOverlayController().installPageOverlay(*findPageOverlay, WebCore::PageOverlay::FadeMode::DoNotFade);
+        findPageOverlay->protectedLayer()->setOpacity(0);
+        m_findPageOverlay = WTFMove(findPageOverlay);
```
Same as above.

RupinMittal commented on Source/WebKit/WebProcess/WebPage/WebFoundTextRangeController.cpp:
```diff
@@ -225,44 +225,46 @@ void WebFoundTextRangeController::clearAllDecoratedFoundText()
-    if (m_findPageOverlay)
-        m_webPage->corePage()->pageOverlayController().uninstallPageOverlay(*m_findPageOverlay, WebCore::PageOverlay::FadeMode::DoNotFade);
+    if (RefPtr findPageOverlay = m_findPageOverlay)
+        m_webPage->protectedCorePage()->pageOverlayController().uninstallPageOverlay(*findPageOverlay, WebCore::PageOverlay::FadeMode::DoNotFade);
```
`pageOverlayController()` is a trivial function, so using just `corePage()` should be ok.

RupinMittal commented on Source/WebKit/WebProcess/WebPage/WebFoundTextRangeController.cpp:
```diff
@@ -442,13 +444,13 @@ void WebFoundTextRangeController::setTextIndicatorWithPDFRange(const WebFoundTex
 {
     Vector<WebCore::FloatRect> rects;
-    Ref mainFrame = m_webPage->corePage()->mainFrame();
+    Ref mainFrame = m_webPage->protectedCorePage()->mainFrame();
```
`mainFrame()` is a trivial function, so just using `corePage()` should be ok.

RupinMittal commented on Source/WebKit/WebProcess/WebPage/WebFoundTextRangeController.cpp:
```diff
@@ -151,10 +151,10 @@ void WebFoundTextRangeController::decorateTextRangeWithStyle(const WebFoundTextR
         switch (style) {
         case FindDecorationStyle::Normal:
-            simpleRange->start.document().markers().removeMarkers(*simpleRange, WebCore::DocumentMarkerType::TextMatch);
+            simpleRange->start.protectedDocument()->checkedMarkers()->removeMarkers(*simpleRange, WebCore::DocumentMarkerType::TextMatch);
```
This is interesting. This seems to be the warning related to this line: https://build.webkit.org/results/Apple-Sequoia-Safer-CPP-Checks/292329@main%20(2568)/scan-build-output/StaticAnalyzer/WebKit/StaticAnalyzerReports/report-WebFoundTextRangeController.cpp-decorateTextRangeWithStyle-15-ebd5f3.html#EndPath

1. The analyzer says `document().markers()` is unsafe. But `markers()` is inline.
2. It doesn't flag `markers().removeMarkers()` as unsafe. But `removeMarkers()` is not trivial. 

Do we want to file bugs for the analyzer here @rniwa?

Also in theory `document()->checkedMarkers()` should work since `checkedMarkers()` is inline. But given that the analyzer may have a bug here within inline functions, what do you recommend @rniwa?

cdumez replied:
@RupinMittal While `markers()` is inline, it is not trivial because it calls a non-inline function: `ensureMarkers()`

RupinMittal reviewed (commented):
A few small comments :)

cdumez commented on Source/WebKit/WebProcess/WebPage/WebFoundTextRangeController.cpp:
```diff
@@ -225,44 +225,46 @@ void WebFoundTextRangeController::clearAllDecoratedFoundText()
     else {
-        m_findPageOverlay = WebCore::PageOverlay::create(*this, WebCore::PageOverlay::OverlayType::Document);
-        m_webPage->corePage()->pageOverlayController().installPageOverlay(*m_findPageOverlay, WebCore::PageOverlay::FadeMode::Fade);
+        RefPtr findPageOverlay = WebCore::PageOverlay::create(*this, WebCore::PageOverlay::OverlayType::Document);
```
Here I would use:
```
Ref findPageOverlay = WebCore::PageOverlay::create(*this, WebCore::PageOverlay::OverlayType::Document);
m_webPage->protectedCorePage()->pageOverlayController().installPageOverlay(findPageOverlay, WebCore::PageOverlay::FadeMode::Fade);
```
`findPageOverlay` cannot be null.

cdumez commented on Source/WebKit/WebProcess/WebPage/WebFoundTextRangeController.cpp:
```diff
@@ -225,44 +225,46 @@ void WebFoundTextRangeController::clearAllDecoratedFoundText()
     }
 
-    m_findPageOverlay->setNeedsDisplay();
+    RefPtr { m_findPageOverlay }->setNeedsDisplay();
```
`RefPtr { m_findPageOverlay }` is used in several places, we should introduce a `protectedFindPageOverlay()` function instead.

cdumez replied:
`RefPtr { m_findPageOverlay }` is used in several places, we should introduce a `protectedFindPageOverlay()` function instead.

cdumez commented on Source/WebKit/WebProcess/WebPage/WebFoundTextRangeController.cpp:
```diff
@@ -225,44 +225,46 @@ void WebFoundTextRangeController::clearAllDecoratedFoundText()
-        m_findPageOverlay = WebCore::PageOverlay::create(*this, WebCore::PageOverlay::OverlayType::Document, WebCore::PageOverlay::AlwaysTileOverlayLayer::Yes);
-        m_webPage->corePage()->pageOverlayController().installPageOverlay(*m_findPageOverlay, WebCore::PageOverlay::FadeMode::DoNotFade);
-        m_findPageOverlay->layer().setOpacity(0);
+        RefPtr findPageOverlay = WebCore::PageOverlay::create(*this, WebCore::PageOverlay::OverlayType::Document, WebCore::PageOverlay::AlwaysTileOverlayLayer::Yes);
```
Can be a `Ref`

cdumez commented on Source/WebKit/WebProcess/WebPage/WebFoundTextRangeController.cpp:
```diff
@@ -413,7 +415,7 @@ void WebFoundTextRangeController::flashTextIndicatorAndUpdateSelectionWithRange(
 
     if (auto textIndicator = createTextIndicatorForRange(range, WebCore::TextIndicatorPresentationTransition::Bounce))
-        m_webPage->setTextIndicator(textIndicator->data());
+        RefPtr { m_webPage.get() }->setTextIndicator(textIndicator->data());
```
Should introduce / use `protectedWebPage()`

cdumez commented on Source/WebKit/WebProcess/WebPage/WebFoundTextRangeController.cpp:
```diff
@@ -442,13 +444,13 @@ void WebFoundTextRangeController::setTextIndicatorWithPDFRange(const WebFoundTex
 {
     if (auto textIndicator = createTextIndicatorForPDFRange(range, WebCore::TextIndicatorPresentationTransition::Bounce))
-        m_webPage->setTextIndicator(textIndicator->data());
+        RefPtr { m_webPage.get() }->setTextIndicator(textIndicator->data());
```
Should introduce / use protectedWebPage()

RupinMittal reviewed (approved):
LGTM. Would be good to wait for the Safer CPP bot since it often catches things we don't.

pxlcoder committed:
Address safer CPP failures in WebFoundTextRangeController.cpp
https://bugs.webkit.org/show_bug.cgi?id=290030
rdar://147385786

Reviewed by Rupin Mittal.

* Source/WebCore/page/PageOverlay.cpp:
(WebCore::PageOverlay::layer const):
(WebCore::PageOverlay::protectedLayer const):
(WebCore::PageOverlay::layer): Deleted.
* Source/WebCore/page/PageOverlay.h:

Drive-by: Make getters const.

* Source/WebCore/page/PageOverlayController.cpp:
(WebCore::PageOverlayController::layerForOverlay const):
* Source/WebCore/page/PageOverlayController.h:
* Source/WebKit/SaferCPPExpectations/UncheckedCallArgsCheckerExpectations:
* Source/WebKit/SaferCPPExpectations/UncountedCallArgsCheckerExpectations:
* Source/WebKit/WebProcess/WebPage/WebFoundTextRangeController.cpp:
(WebKit::WebFoundTextRangeController::findTextRangesForStringMatches):
(WebKit::WebFoundTextRangeController::replaceFoundTextRangeWithString):
(WebKit::WebFoundTextRangeController::decorateTextRangeWithStyle):
(WebKit::WebFoundTextRangeController::clearAllDecoratedFoundText):
(WebKit::WebFoundTextRangeController::didBeginTextSearchOperation):
(WebKit::WebFoundTextRangeController::addLayerForFindOverlay):
(WebKit::WebFoundTextRangeController::removeLayerForFindOverlay):
(WebKit::WebFoundTextRangeController::createTextIndicatorForRange):
(WebKit::WebFoundTextRangeController::flashTextIndicatorAndUpdateSelectionWithRange):
(WebKit::WebFoundTextRangeController::flashTextIndicatorAndUpdateSelectionWithPDFRange):
(WebKit::WebFoundTextRangeController::rectsForTextMatchesInRect):
(WebKit::WebFoundTextRangeController::frameForFoundTextRange const):

Canonical link: https://commits.webkit.org/292430@main
